### ISSUE TITLE
performance issues

### ISSUE THREAD

**Issue by sglyon** [EconForge/Dolang#3]
TL;DR `_parse` is slow and is called more than it should. derivatives using Calculus.jl are by far the slowest part of the code and probably not a feasible long run solution.

I had  Dolo.jl load the [EA_QUEST3.mod](https://github.com/EconForge/Dolo.jl/blob/f13eb0c7e7ff77d2e297c8591fb7953cb2e8d744/examples/models/dynare/EA_QUEST3.mod) file and compile up to the second order derivatives. It took about 20 seconds. The corresponding benchmark for the dynare preprocessor is under 1.5 seconds. 

I profiled the `dynare_import` function and saw that we spend almost all our time in this routine. Looking through the profile results, I noticed the following:
- we spend _a lot_ of time computing derivatives with Calculus.jl. I don't think it is a good long run solution -- we should look for something else.
- Inside Dolang we have many extra calls to `_parse` this is expensive. We should eliminate them. The worse case is normalizing argument names in the nested loop where we compute hessian expressions. This is totally un-necessary. We should only have to normalize them once -- ever.
- The `_parse` function itself is not very well written. The slow part is the pattern matching, but 99% of the time we don't need it. We often look for specific forms of `Expr`, which we could easily do on our own.

**Comment by spencerlyon2** [comment]
I ripped out the pattern matching from `_parse` and am now doing it by hand. It is much much faster now. Calling `map(_parse, ea_quest3_equations)` is down from 1.05 seconds to 0.002 seconds.

Constructing the derivatives is also much faster now that I am utilizing information in the incidence table to avoid having Calculus.jl compute a bunch of 0 derivatives.

Right now we are at about 4.5 seconds for loading the model (down from 20).

**Comment by albop** [comment]
Just to be clear, this includes derivatives up to order 1 (or 2) ?

**Comment by spencerlyon2** [comment]
So the latest update is that we are down to about 3.1 seconds for starting to parse the module to compiling functions; including levels, first derivatives, and second derivatives. 

I did some more profiling of the code that ran in 3.1 seconds and learned the following (remember, Julia uses an sampling profiler so the output is not "the code spend x% of its time on this line"):

- 611 out of 2597 total hits were in `solve_triangular_system` when trying to construct the calibration. All 611 of these were on the call to `eval`. We can implement a more efficient algorithm that does incidence analysis like what is in dolo python to drastically bring this number down.
- The most costly part was computing the derivatives when trying to build the Hessian. Differentiating wrt v1 was 148/2597 and wrt v2 was 659/2597. All but 6 of those hits spent their time in Calculus.jl (so 801/2597 are in Calculus.jl). It might be worthwhile to benchmark another option for derivatives (probably symengine). 
- Buliding the `FunctionFactory` from the symbolic model took 201/2597, but that actually does a ton of work (buliding incidence, filtering args, validating equations). That doesn't seem like it sucks too much time.
- At least 122/2597 spent `eval`ing the generated code (i.e. making the functions callable). That's mostly out of our hands.

**Comment by albop** [comment]
Just to get some orders of magnitude : I did a quick experiment in python
yesterday. It takes 1/2 second to parse and normalize equations by using
the ast module.
Computing Jacobian and Hessian w.r.t. to all variables using
python/symengine takes ... 0.10 second !
But sympifying the equations and sending them to symengine takes a few
seconds (3-4). I haven't found a way to parse them directly with symengine.
I'll upload the code soon but the bottom line is that symengine looks like
a very good option (albeit it might comes with installation problems).
As for solve_triangular_system, we should aim to have smthg almost
instantaneous.

**Comment by spencerlyon2** [comment]
0.10 is very fast! That's great to hear that we have a faster option out there.

So I benchmarked each main piece of the routine by itself and these are the timings:

```
regex_create_symbolic_model:
BenchmarkTools.Trial:
  samples:          35
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.34 mb
  allocs estimate:  168470
  minimum time:     137.51 ms (0.00% GC)
  median time:      147.47 ms (0.00% GC)
  mean time:        146.54 ms (1.05% GC)
  maximum time:     153.32 ms (3.09% GC)


parse_normalize:
BenchmarkTools.Trial:
  samples:          2944
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  524.91 kb
  allocs estimate:  11771
  minimum time:     1.45 ms (0.00% GC)
  median time:      1.55 ms (0.00% GC)
  mean time:        1.70 ms (4.60% GC)
  maximum time:     6.61 ms (61.71% GC)


build_function_factory:
BenchmarkTools.Trial:
  samples:          604
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  3.32 mb
  allocs estimate:  66333
  minimum time:     6.70 ms (0.00% GC)
  median time:      7.23 ms (0.00% GC)
  mean time:        8.27 ms (7.35% GC)
  maximum time:     20.04 ms (48.94% GC)


build_levels_func_body:
BenchmarkTools.Trial:
  samples:          8748
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  269.05 kb
  allocs estimate:  4942
  minimum time:     416.88 μs (0.00% GC)
  median time:      477.78 μs (0.00% GC)
  mean time:        567.82 μs (8.56% GC)
  maximum time:     7.40 ms (87.84% GC)


compute_jacobian_derivatives:
BenchmarkTools.Trial:
  samples:          27
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  37.34 mb
  allocs estimate:  325550
  minimum time:     161.18 ms (2.49% GC)
  median time:      189.88 ms (3.75% GC)
  mean time:        187.71 ms (3.59% GC)
  maximum time:     210.62 ms (3.59% GC)


build_jacobian_func_body:
BenchmarkTools.Trial:
  samples:          28
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  37.68 mb
  allocs estimate:  332246
  minimum time:     161.76 ms (2.45% GC)
  median time:      179.73 ms (3.88% GC)
  mean time:        179.44 ms (3.73% GC)
  maximum time:     193.81 ms (4.62% GC)


compute_hessian_derivatives:
BenchmarkTools.Trial:
  samples:          5
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  230.12 mb
  allocs estimate:  2030341
  minimum time:     989.20 ms (4.23% GC)
  median time:      1.03 s (4.93% GC)
  mean time:        1.05 s (4.78% GC)
  maximum time:     1.12 s (4.82% GC)


build_hessian_func_body:
BenchmarkTools.Trial:
  samples:          5
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  232.03 mb
  allocs estimate:  2057923
  minimum time:     1.04 s (4.72% GC)
  median time:      1.05 s (4.72% GC)
  mean time:        1.06 s (4.75% GC)
  maximum time:     1.11 s (3.55% GC)
```

Each of the "keys" above the `BenchmarkTools.Trial` results do the following:

- `regex_create_symbolic_model`: use code from @jagman88 to parse the mod file
using regex and build the SymbolicModel
- `parse_normalize`: Parse and normalize all the equations in the model
- `build_function_factory`: Build the funciton factory from the `SymbolicModel`.
This includes building the incidence tables, validating equations, and filtering
the arguments
- `build_levels_func_body`: Use the `FunctionFactory` to construct the body of
the function that evaluates the levels of the function for the model
- `compute_jacobian_derivatives`: Compute all non-zero derivatives used in
constructing the jacobian functino
- `build_jacobian_func_body`: Use the `FunctionFactory` to construct the body of
the function that evaluates the Jacobian of the function for the model. NOTE:
this also includes time to compute the derivatives, so that time appears here
and in the previous set of results
- `compute_hessian_derivatives`: Compute all non-zero derivatives used in building
the hessian function
- `build_hessian_func_body`: Use the `FunctionFactory` to construct the body of
the function that evaluates the hessian of the function for the model. NOTE:
this also includes time to compute the derivatives, so that time appears here
and in the previous set of results

From the results it seems fairly clear that we are spending the majority of our time computing derivatives (especially for the Hessian) using Calculus.jl.

I'll see what I can do today to fix up the SymEngine.jl bindings so they are usable for us.

**Comment by spencerlyon2** [comment]
Oh, and I forgot a benchmark for `solve_triangular_system`. This function just takes the calibration inside the `SymbolicModel` and converts it to all numerical values. Here are the benchmark results:

```
BenchmarkTools.Trial:
  samples:          7
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  18.14 mb
  allocs estimate:  337525
  minimum time:     806.35 ms (0.37% GC)
  median time:      817.20 ms (0.36% GC)
  mean time:        820.14 ms (0.28% GC)
  maximum time:     838.72 ms (0.42% GC)
```

**Comment by spencerlyon2** [comment]
I have repeated the benchmarks from above now that we have the SymEngine backend working. Here are the numbers:

```
julia> for k in keys(bench.data)
       println("Benchmark for $k:")
       show(bench[k])
       println("\n\n")
       end
Benchmark for build_levels_func_body:
BenchmarkTools.Trial:
  samples:          8955
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  277.17 kb
  allocs estimate:  4695
  minimum time:     390.44 μs (0.00% GC)
  median time:      438.19 μs (0.00% GC)
  mean time:        556.77 μs (14.26% GC)
  maximum time:     13.56 ms (0.00% GC)


Benchmark for build_jacobian_func_body:
BenchmarkTools.Trial:
  samples:          169
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  1.25 mb
  allocs estimate:  25317
  minimum time:     24.84 ms (0.00% GC)
  median time:      28.22 ms (0.00% GC)
  mean time:        29.64 ms (1.61% GC)
  maximum time:     60.00 ms (0.00% GC)


Benchmark for regex_create_symbolic_model:
BenchmarkTools.Trial:
  samples:          37
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  3.01 mb
  allocs estimate:  62834
  minimum time:     129.21 ms (0.00% GC)
  median time:      135.67 ms (0.00% GC)
  mean time:        136.42 ms (0.78% GC)
  maximum time:     151.87 ms (4.94% GC)


Benchmark for compute_jacobian_derivatives:
BenchmarkTools.Trial:
  samples:          186
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  905.03 kb
  allocs estimate:  18535
  minimum time:     23.71 ms (0.00% GC)
  median time:      24.89 ms (0.00% GC)
  mean time:        26.95 ms (1.17% GC)
  maximum time:     57.22 ms (14.53% GC)


Benchmark for build_hessian_func_body:
BenchmarkTools.Trial:
  samples:          63
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  2.57 mb
  allocs estimate:  56372
  minimum time:     71.68 ms (0.00% GC)
  median time:      77.03 ms (0.00% GC)
  mean time:        80.39 ms (1.49% GC)
  maximum time:     115.18 ms (9.93% GC)


Benchmark for parse_normalize:
BenchmarkTools.Trial:
  samples:          2210
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  525.50 kb
  allocs estimate:  9603
  minimum time:     1.72 ms (0.00% GC)
  median time:      1.93 ms (0.00% GC)
  mean time:        2.26 ms (7.18% GC)
  maximum time:     46.90 ms (0.00% GC)


Benchmark for build_function_factory:
BenchmarkTools.Trial:
  samples:          357
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  5.09 mb
  allocs estimate:  82546
  minimum time:     11.08 ms (0.00% GC)
  median time:      12.33 ms (0.00% GC)
  mean time:        14.05 ms (10.94% GC)
  maximum time:     29.54 ms (28.73% GC)


Benchmark for compute_hessian_derivatives:
BenchmarkTools.Trial:
  samples:          61
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  1.65 mb
  allocs estimate:  41447
  minimum time:     69.71 ms (0.00% GC)
  median time:      77.63 ms (0.00% GC)
  mean time:        82.58 ms (3.54% GC)
  maximum time:     301.92 ms (51.12% GC)


Benchmark for solve_triangular_system:
BenchmarkTools.Trial:
  samples:          6
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  21.94 mb
  allocs estimate:  382939
  minimum time:     793.43 ms (0.00% GC)
  median time:      862.25 ms (0.00% GC)
  mean time:        994.44 ms (4.30% GC)
  maximum time:     1.40 s (0.79% GC)
```

The main takeaway is that solve_triangular_system is low hanging fruit ripe for a performance tune up

**Comment by spencerlyon2** [comment]
Ok I re-implemented the triangular solver and got a 100x speedup for that part.

This impacts the `dynare_import` benchmark as well as the `solve_triangular_system` benchmark. The updated times are below:

```
Benchmark for solve_triangular_system:
BenchmarkTools.Trial:
  samples:          1195
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  1.03 mb
  allocs estimate:  6571
  minimum time:     7.14 ms (0.00% GC)
  median time:      7.80 ms (0.00% GC)
  mean time:        8.37 ms (4.88% GC)
  maximum time:     152.11 ms (95.08% GC)


Benchmark for dynare_import:
BenchmarkTools.Trial:
  samples:          10
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  26.98 mb
  allocs estimate:  579614
  minimum time:     985.37 ms (0.93% GC)
  median time:      1.01 s (1.03% GC)
  mean time:        1.01 s (1.23% GC)
  maximum time:     1.03 s (0.99% GC)
```

We're getting close to a reasonable place!

**Comment by albop** [comment]
Great ! What is the new bottleneck ?

**Comment by spencerlyon2** [comment]
The latest version of the whole benchmark suite is in this gist: https://gist.github.com/spencerlyon2/af9d604259ab77647edecf8331fe4d16

It looks like the string parsing using regex is the slowest part by about a factor of 2. After that it is taking all the 2nd order derivatives and then building the expressions for the body of the second order derivative function.
